core/trace/dt_span_serializer.go

```diff
@@ -75,13 +80,93 @@ func serializeSpans(
 	}
 
 	spanExport := &protoCollectorTraces.SpanExport{
-		TenantUUID:     tenantUUID,
-		AgentId:        agentId,
+		TenantUUID:     s.tenantUUID,
+		AgentId:        s.agentId,
 		ExportMetaInfo: exportMetaInfo,
 		Resource:       serializedResource,
-		Spans:          agSpanEnvelopes,
 	}
-	return proto.Marshal(spanExport)
+
+	exports := []*protoCollectorTraces.SpanExport{spanExport}
+
+	spanlessMsgSize := proto.Size(spanExport)
+
+	s.logger.Debugf("spanless message size: %v", spanlessMsgSize)
+
+	if spanlessMsgSize > cMsgSizeMax {
+		return nil, fmt.Errorf("resource too big (%v), cannot export any spans", spanlessMsgSize)
+	}
+
+	sizeSoFar := spanlessMsgSize
+
+	agSpanEnvelopes := make([]*protoCollectorTraces.ActiveGateSpanEnvelope, 0, len(spans))
+
+	for span := range spans {
+		fw4Tag := span.metadata.fw4Tag
+		customTag := getProtoCustomTag(fw4Tag.CustomBlob)
+
+		spanMsg, err := createProtoSpan(span, customTag, s.qualifiedTenantId)
+		if err != nil {
+			return nil, err
+		}
+
+		estimatedEnvelopeSize := 0
+
+		serializedClusterSpanEnvelope, err := createSerializedClusterSpanEnvelope(spanMsg, customTag, int32(fw4Tag.PathInfo))
+		if err != nil {
+			return nil, err
+		}
+
+		// Estimate 1 byte for the tag size and up to 4 byte for the varint
+		// encoding the size of the span container.
```

Comment:
I believe this estimation could be replaced with a call to https://pkg.go.dev/google.golang.org/protobuf@v1.31.0/encoding/protowire#SizeGroup